src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -42,9 +43,11 @@ internal sealed class TensorFlowMapper : IRowMapper
             private readonly TFShape[] _tfInputShapes;
             private readonly TFDataType[] _tfInputTypes;
 
-            private readonly string _outputColName;
-            private readonly ColumnType _outputColType;
-            private readonly TFDataType _tfOutputType;
+            private readonly string[] _outputColNames;
+            private readonly ColumnType[] _outputColTypes;
+            private readonly TFDataType[] _tfOutputTypes;
+            private IDictionary<string, TFTensor> _cachedOutputs;
+            private long _cachedPosition;
```

Comment:
question: what do the _cachedOutputs and _cachedPosition represent ?  

Are they only needed for the multiple inputs/outputs scenario ?

Reply:
Actually, output getters work independent of each other and in each getter we have call to compute operations on TF graph. 

What we are doing here is for each row in the input dataview, cache the results of all the requested operations on the first output getter call. Then for subsequent output getter calls for that row, serve the results from this `_cachedOutputs` dictionary.

If the row position is changed the computation is done again and the current position is saved in `cachedPosition`.

---
In reply to: [215393239](https://github.com/dotnet/machinelearning/pull/814#discussion_r215393239) [](ancestors = 215393239)